File: results/RQ1/05_baseline_qualification/units/p04_caro_game/java/accepted/repository/src/main/java/Board.java
/**
 * Represents the game board as a square grid.
 * The board stores player tokens with null representing empty cells.
 */
public final class Board {
    private final int size;
    private final Player[][] grid;

    /**
     * Creates a new Board with the specified size.
     *
     * @param size the size of the square board (must be >= 5)
     * @throws IllegalArgumentException if size < 5
     */
    public Board(int size) {
        if (size < 5) {
            throw new IllegalArgumentException("Board size must be at least 5");
        }
        this.size = size;
        this.grid = new Player[size][size];
    }

    /**
     * Returns the size of the board.
     *
     * @return the board size
     */
    public int getSize() {
        return size;
    }

    /**
     * Checks if the given coordinates are within the board boundaries.
     *
     * @param row the row coordinate
     * @param col the column coordinate
     * @return true if the coordinates are valid, false otherwise
     */
    public boolean isInBounds(int row, int col) {
        return row >= 0 && row < size && col >= 0 && col < size;
    }

    /**
     * Gets the player at the specified position.
     *
     * @param row the row coordinate
     * @param col the column coordinate
     * @return the player at the position, or null if empty
     * @throws IllegalArgumentException if coordinates are out of bounds
     */
    public Player getCell(int row, int col) {
        if (!isInBounds(row, col)) {
            throw new IllegalArgumentException("Coordinates out of bounds");
        }
        return grid[row][col];
    }

    /**
     * Checks if the cell at the specified position is empty.
     *
     * @param row the row coordinate
     * @param col the column coordinate
     * @return true if the cell is empty, false otherwise
     * @throws IllegalArgumentException if coordinates are out of bounds
     */
    public boolean isEmpty(int row, int col) {
        if (!isInBounds(row, col)) {
            throw new IllegalArgumentException("Coordinates out of bounds");
        }
        return grid[row][col] == null;
    }

    /**
     * Sets the cell at the specified position to the given player.
     * The operation only succeeds if the cell is empty and coordinates are valid.
     *
     * @param row the row coordinate
     * @param col the column coordinate
     * @param player the player to set
     * @return true if the cell was successfully set, false otherwise
     * @throws IllegalArgumentException if coordinates are out of bounds
     */
    public boolean setCell(int row, int col, Player player) {
        if (!isInBounds(row, col)) {
            throw new IllegalArgumentException("Coordinates out of bounds");
        }
        if (grid[row][col] != null) {
            return false;
        }
        grid[row][col] = player;
        return true;
    }

    /**
     * Clears the cell at the specified position.
     * The operation only succeeds if the coordinates are valid.
     *
     * @param row the row coordinate
     * @param col the column coordinate
     * @throws IllegalArgumentException if coordinates are out of bounds
     */
    public void clearCell(int row, int col) {
        if (!isInBounds(row, col)) {
            throw new IllegalArgumentException("Coordinates out of bounds");
        }
        grid[row][col] = null;
    }

    /**
     * Returns a defensive copy of the board's current state.
     * Each element in the returned array is a reference to the corresponding
     * player on the board (null for empty).
     *
     * @return a 2D array representing the board state
     */
    public Player[][] snapshot() {
        Player[][] snapshot = new Player[size][size];
        for (int i = 0; i < size; i++) {
            System.arraycopy(grid[i], 0, snapshot[i], 0, size);
        }
        return snapshot;
    }
}

File: results/RQ1/05_baseline_qualification/units/p04_caro_game/python/replays/0001_seed_812711061/p04_caro_game_python/p04_caro_game_python/caro/board.py
"""Board representation and operations for the Caro game."""

from typing import List, Optional
from .types import Player


class Board:
    """Represents the game board for Caro.

    The board is a square grid of a specified size, initialized with empty cells.
    Each cell can contain a Player (X or O) or be empty (None).
    """

    def __init__(self, size: int = 15) -> None:
        """Initialize the board with a given size.

        Args:
            size: The dimension of the square board. Must be at least 1.
        """
        if size < 1:
            raise ValueError("Board size must be at least 1")
        self.size = size
        # Initialize grid with None values representing empty cells
        self.grid: List[List[Optional[Player]]] = [
            [None for _ in range(size)] for _ in range(size)
        ]

    def in_bounds(self, row: int, col: int) -> bool:
        """Check if the given coordinates are within the board boundaries.

        Args:
            row: The row index to check.
            col: The column index to check.

        Returns:
            True if both row and col are within valid bounds, False otherwise.
        """
        return 0 <= row < self.size and 0 <= col < self.size

    def get_cell(self, row: int, col: int) -> Optional[Player]:
        """Get the player occupying a specific cell.

        Precondition: The coordinates must be in bounds.
        This method assumes the caller has verified the coordinates are valid.

        Args:
            row: The row index of the cell.
            col: The column index of the cell.

        Returns:
            The player who occupies the cell, or None if it's empty.
        """
        return self.grid[row][col]

    def is_empty(self, row: int, col: int) -> bool:
        """Check if a specific cell is empty.

        Precondition: The coordinates must be in bounds.
        This method assumes the caller has verified the coordinates are valid.

        Args:
            row: The row index of the cell.
            col: The column index of the cell.

        Returns:
            True if the cell is empty (None), False otherwise.
        """
        return self.grid[row][col] is None

    def set_cell(self, row: int, col: int, player: Player) -> None:
        """Set a player in a specific cell.

        Precondition: The coordinates must be in bounds and the cell must be empty.
        This method assumes the caller has verified these conditions.

        Args:
            row: The row index of the cell.
            col: The column index of the cell.
            player: The player to place in the cell.
        """
        self.grid[row][col] = player

    def is_full(self) -> bool:
        """Check if the board is completely filled with players.

        Returns:
            True if all cells are occupied, False otherwise.
        """
        for row in self.grid:
            if None in row:
                return False
        return True

    def to_matrix(self) -> List[List[Optional[Player]]]:
        """Create a deep copy of the board's grid.

        Returns:
            A new 2D list representing the current state of the board.
        """
        # Create a deep copy of the grid
        return [row[:] for row in self.grid]